### PR TITLE
RSDK-4196 Only allow flat tensors in mlmodel

### PR DESCRIPTION
**Pull request by benjirewis**
[RSDK-4196](https://viam.atlassian.net/browse/RSDK-4196)

https://github.com/viamrobotics/viam-cpp-sdk/pull/122 added support for flat tensors per the new proto definition but still allowed the old input and output paths. A recent proto change removed the old input and output paths from the API, so only flat tensors are allowed. Updates the CPP SDK accordingly.

[RSDK-4196]: https://viam.atlassian.net/browse/RSDK-4196?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/viam/sdk/services/mlmodel/client.cpp [hunk 1]
```diff
@@ -45,77 +39,36 @@ std::shared_ptr<MLModelService::named_tensor_views> MLModelServiceClient::infer(
...
-
-        auto& mutable_input_data = *req->mutable_input_data();
-        auto& mutable_input_data_fields = *mutable_input_data.mutable_fields();
-
-        // TODO: Currently, this doesn't validate that we are passing the
-        // right input type in. We could query the metadata here and
-        // consult it.
-        for (const auto& kv : inputs) {
-            pb::Value& value = mutable_input_data_fields[kv.first];
-            mlmodel_details::tensor_to_pb_value(kv.second, &value);
```
Comment (acmorrow) [inline src/viam/sdk/services/mlmodel/client.cpp:63]: Please remove the declaration and implementation (and associated helper functions/types) of `mlmodel_details::tensor_to_pb_value`.

File: src/viam/sdk/services/mlmodel/client.cpp [hunk 2]
```diff
@@ -45,77 +39,36 @@ std::shared_ptr<MLModelService::named_tensor_views> MLModelServiceClient::infer(
...
-        // it is not clear that caching it is safe.
-        const auto md = metadata();
-
-        const auto& output_fields = resp->output_data().fields();
-        for (const auto& output : md.outputs) {
-            const auto where = output_fields.find(output.name);
-            // Ignore any outputs for which we don't have metadata, since
-            // we can't know what type they should decode to.
-            if (where != output_fields.end()) {
-                mlmodel_details::pb_value_to_tensor(
```
Comment (acmorrow) [inline src/viam/sdk/services/mlmodel/client.cpp:81]: Ditto `mlmodel_details::pb_value_to_tensor` and associated helpers.

File: src/viam/sdk/services/mlmodel/server.cpp [hunk 3]
```diff
@@ -51,11 +51,7 @@ ::grpc::Status MLModelServiceServer::Infer(
 
     std::shared_ptr<MLModelService> mlms = std::dynamic_pointer_cast<MLModelService>(rb);
 
-    if (request->has_input_data() && request->has_input_tensors()) {
-        return {::grpc::StatusCode::INVALID_ARGUMENT, "Called [Infer] with both forms of input"};
-    }
-
-    if (!request->has_input_data() && !request->has_input_tensors()) {
+    if (!!request->has_input_tensors()) {
```
Comment (acmorrow) [inline src/viam/sdk/services/mlmodel/server.cpp:54]: I think there is an extra `!` here.

File: src/viam/sdk/services/mlmodel/server.cpp [hunk 4]
```diff
@@ -64,92 +60,39 @@ ::grpc::Status MLModelServiceServer::Infer(
...
-        const auto md = mlms->metadata(extra);
-
-        mlmodel_details::tensor_storage input_storage;
-        MLModelService::named_tensor_views inputs;
-
-        const auto& input_fields = request->input_data().fields();
-        for (const auto& input : md.inputs) {
-            const auto where = input_fields.find(input.name);
-
+    const auto md = mlms->metadata(extra);
```
Comment (acmorrow) [inline src/viam/sdk/services/mlmodel/server.cpp:63]: Git didn't do a great job with this diff. Is it safe to assume that this is simply a removal of the `has_input_data` path with no meaningful changes to the `has_input_tensors` path?

File: src/viam/sdk/services/mlmodel/client.cpp [hunk 5]
```diff
@@ -45,77 +39,36 @@ std::shared_ptr<MLModelService::named_tensor_views> MLModelServiceClient::infer(
...
-        for (const auto& kv : resp->output_tensors().tensors()) {
-            // NOTE: We don't need to pass in tensor storage here,
-            // because the backing store for the views is the Arena we
-            // moved into our result above.
-            auto tensor = mlmodel_details::make_sdk_tensor_from_api_tensor(kv.second);
-            aav->views.emplace(kv.first, std::move(tensor));
-        }
-        auto* const tsav_views = &aav->views;
-        return {std::move(aav), tsav_views};
+    struct arena_and_views {
```
Comment (acmorrow) [inline src/viam/sdk/services/mlmodel/client.cpp:42]: Git didn't do a great job with this diff. Safe to assume this is just a removal of the `!kUseFlatTensors` path with no changes to the `kUseFlatTensors` code?

**Comment by benjirewis** [comment]
Sorry for early review request; still some test failures.

**Review by acmorrow — CHANGES_REQUESTED** [review]

File: src/viam/sdk/services/mlmodel/client.cpp (hunk 2 above)
Comment (benjirewis) [reply src/viam/sdk/services/mlmodel/client.cpp:81]: Can do.

File: src/viam/sdk/services/mlmodel/client.cpp (hunk 5 above)
Comment (benjirewis) [reply src/viam/sdk/services/mlmodel/client.cpp:42]: Yep!

File: src/viam/sdk/services/mlmodel/server.cpp (hunk 3 above)
Comment (benjirewis) [reply src/viam/sdk/services/mlmodel/server.cpp:54]: ...coding sans coffee. Thanks! This was responsible for the test failure.

File: src/viam/sdk/services/mlmodel/server.cpp (hunk 4 above)
Comment (benjirewis) [reply src/viam/sdk/services/mlmodel/server.cpp:63]: Yep!

**Review by acmorrow — APPROVED** [review]
LGTM if tests are passing now

**Review by stuqdog — APPROVED** [review]
lgtm!

**Pull request merged**